### ISSUE TITLE
Ignore PROJECT.JSON when using .CSPROJ

### ISSUE THREAD

**Issue by bchavez** [Microsoft/msbuild#394]
Hi,

I have a serious problem after upgrading to VS 2015 Update 1 RTM.

I am trying to do cross-platform development for this [RethinkDB Driver](https://github.com/bchavez/RethinkDb.Driver/tree/master/Source/RethinkDb.Driver) for both CoreCLR/DNX and full .NET 4.5 framework.

I don't want to convert everything to a DNX build project because the CoreCLR/DNX stuff has not RTM'd. So, I'd just like to use standard RTM tools when developing the .NET 4.5 target, then, later when Visual Studio is closed, do `dnu build` ***as a separate thing***.

When I have  **.CSPROJ** open in Visual Studio and I try to compile, I get an error that was not here prior to **Update 1**:

```
Your project.json doesn't have a runtimes section. 
You should add '"runtimes": { "win": { } }' to your project.json and 
then re-run NuGet restore. RethinkDb.Driver
```

Okay... So I attempt to add:
```
"runtimes": { "win": { } }
```
**Q1**: why do I need this "win" tag since I'm doing cross-platform targeting with CoreCLR? I am not, **in any** way, targeting "windows" with CoreCLR/DNX).

So, I place the tag inside **project.json**, but still no dice. Now I'm getting even more errors:

```
Some packages are not compatible with DNXCore,Version=v5.0 (win).			0	
System.Runtime.Extensions 4.0.11-beta-23516 provides a compile-time reference assembly for System.Runtime.Extensions on DNXCore,Version=v5.0, but there is no run-time assembly compatible with win.			0	
System.Net.Primitives 4.0.11-beta-23516 provides a compile-time reference assembly for System.Net.Primitives on DNXCore,Version=v5.0, but there is no run-time assembly compatible with win.			0	
```

Now it looks like this new "win" runtime tag is causing even more problems. Can someone tell me why MSBUILD is so unhappy now?

Even better, can someone tell me how to force ***MSBUILD + VS + NUGET*** to :no_entry_sign: **STOP** :no_entry_sign: probing for **project.json** when I have my classic **.CSPROJ** open? I ***do not***, want **project.json** even being considered or looked at when I have my classic **.CSPROJ** open in Visual Studio. I'd prefer my **PACKAGES.CONFIG** :white_check_mark: considered when my classic project is open. I don't understand why the tooling from both DNX (project.json) and CSPROJ are conflicting. IMHO, these should be two **completely** different things at this stage.

Feel free to fork or download the project. My sources here: https://github.com/bchavez/RethinkDb.Driver

There are two solutions: 
* `RethinkDb.Driver.Dnx.sln` - DNX Project Solution
* `RethinkDb.Driver.sln` - Classic **.csproj** Solution

Thanks,
Brian

**Comment by bchavez** [comment]
Okay, I *sort of* solved this by downgrading **.csproj** `Toolsversion` to **4.0** (from **14.0**) `<Project ToolsVersion="4.0"`.

**Comment by weshaggard** [comment]
cc @Pilchie @jasonmalinowski

**Comment by Pilchie** [comment]
`project.json` and .csproj files are *NOT* two differently things.  `project.json` is used by more than just DNX.  It can also be used in conjunction with .csproj files as a replacement for packages.config, to allow .csproj files to take advantage of features like transitive dependencies, build time resolution of assets, etc.

The reason you are being prompted to add a runtime section is because the build time asset resolution is trying to copy assets from your dependencies to allow you to run, but it needs to know what assets you want to copy.

Having a project.json meant to be used for DNX in the same directory as a packages.config to be used by msbuild isn't something that's supported today, but you *can* have two different project.json files, one for msbuild and one for DNX.  To do that, create a "<projectname>.project.json" (where <projectname> is the name of your .csproj.  This will allow you to specify the existing framework from your .csproj, instead of of DNX.

Finally, if you are building a library and you *don't* want to copy runtime assets to your output directory, you can add <CopyNuGetImplementations>false</CopyNuGetImplementations> to your .csproj, in which case you won't need a `runtimes` section in your <projectname>.project.json.

Hope this helps.

(Also tagging @yishaigalatzer and @ericstj)

**Comment by bchavez** [comment]
Thanks @Pilchie 

>be used in conjunction with .csproj files as a replacement for packages.config

Thank for explaining this: I'll migrate off **packages.config**

>The reason you are being prompted to add a runtime section is because the build time asset resolution is trying to copy assets from your dependencies to allow you to run, but it needs to know what assets you want to copy.

If I understand correctly, then, how do I tell the **build time asset resolution** that I only want outputs for `net45` full framework right now when I build the solution in VS2015 and not bother with any DNX stuff?

>if you are building a library and you don't want to copy runtime assets to your output directory, you can add false to your .csproj

Add false where?

Again, thanks for your help

**Comment by bchavez** [comment]
Maybe, specify `runtime: { "win":{}} `  in **_name_.project.json**, but leave it out in the root `project.json`?

**Comment by Pilchie** [comment]
I edited my markdown so that the xml doesn't disappear in in the rendered output.

Right - also, *name.project.json* should just have a single entry like `net451` in the frameworks section.

**Comment by bchavez** [comment]
Fantastic. Thank you. That seems to have solved everything. Once again, happily compiling! I'm sure other will find this information *really* useful.

**Comment by bchavez** [comment]
This is a bit strange. So, check this out:

1. Everything was working fine when I closed the issue. I could compile fine in VS2015 Update 1 after adding **name.project.json**.
2. I added one dependency on `System.Threading.Timer` to `project.json`:
```
"dnxcore50": {
      "dependencies": {
        //
        "System.Threading.Timer": "4.0.1-beta-23516"
      },
```
3. Build, but got the error: 
```
Your project.json doesn't have a runtimes section. 
You should add '"runtimes": { "win": { } }' to your project.json and 
then re-run NuGet restore. RethinkDb.Driver
```
4. Ok. So, I back out and remove the line `System.Threading.Timer`, save, build-clean, and "Restored NuGet" packages. Even restarted Visual Studio and a ***still*** get an error message.

So, now ***what*** is going on? I have `name.project.json` with the necessary "runtime", and a `project.json` without any runtimes as we discussed few days ago. 

I dunno, but to me it seems natural to simply reverting my 1 line edit should get me back to where I was but I cannot back out of a 1-line edit. I cannot get rid of this error. :-1:. Back to square one in this original issue.

Latest code, "name.project.json" & "project.json" files here:
https://github.com/bchavez/RethinkDb.Driver/tree/master/Source/RethinkDb.Driver

CI server's Build Error:
https://ci.appveyor.com/project/bchavez/rethinkdb-driver/build/282#L654

```
"C:\projects\rethinkdb-driver\Source\RethinkDb.Driver.sln" (Rebuild target) (1) ->
"C:\projects\rethinkdb-driver\Source\RethinkDb.Driver.Tests\RethinkDb.Driver.Tests.csproj" (Rebuild target) (2) ->
"C:\projects\rethinkdb-driver\Source\RethinkDb.Driver\RethinkDb.Driver.csproj" (default target) (3:2) ->
(ResolveNuGetPackageAssets target) -> 
  C:\Program Files (x86)\MSBuild\Microsoft\NuGet\Microsoft.NuGet.targets(109,5): error : Your project.json doesn't list 'win' as a targeted runtime. You should add '"win": { }' inside your "runtimes" section in your project.json, and then re-run NuGet restore. [C:\projects\rethinkdb-driver\Source\RethinkDb.Driver\RethinkDb.Driver.csproj]
 
    0 Warning(s)
    1 Error(s)
 
Time Elapsed 00:00:04.14
```

**Comment by bchavez** [comment]
I deleted these `*.json.lock` files and it seems to build again. I still think this is a symptom of a larger problem.

1. I delete ***ALL*** lock files (and ensure NO lock files exist in the build DIR), I can run

    `msbuild source\RethinkDb.Driver.sln`, ***BUILD SUCCESS***

2. **Restore Packages** from Visual Studio 2015 Update 1, generates `RethinkDb.Driver.project.lock.json`,  and the command:

    `msbuild source\RethinkDb.Driver.sln` ***BUILD FAIL*** with error message below:

```
Your project.json doesn't list 'win' as a targeted runtime. You should add '"win": { }' 
inside your "runtimes" section in your project.json, and then re-run NuGet restore.
	RethinkDb.Driver			
```
So, ironically, **Restore Packages** breaks the build. :8ball:

**Comment by jasonmalinowski** [comment]
So there's a two things there:

1. I've filed NuGet/Home#1859 to address your immediate point that we're using the existence of the lock file to figure out whether we are looking for NuGet packages or not. In your case, the first build _should_ fail with some sort of "you didn't restore packages" error. In your case, the build succeeded but didn't actually consume the packages you thought you referenced, so that isn't good.
2. The error message is correct: you need to update your project.json to include that entry under runtimes. In Visual Studio 2015 RTM there was a bug that meant you could get away with win-anycpu in there, but that had many other problems. You should add the win entry, and remove win-anycpu if you don't care about building with RTM.

**Comment by bchavez** [comment]
Hi @jasonmalinowski,

I have done everything @Pilchie has told me to do. As far as I understand, regarding your No.2 suggestion, it seems your advice is going *against* what @Pilchie  suggested I do regarding `name.project.json` that only specifies a "runtimes" property.

1. I have [`project.json`](https://github.com/bchavez/RethinkDb.Driver/blob/master/Source/RethinkDb.Driver/project.json) with all the necessary references.
2. I have [`name.project.json`](https://github.com/bchavez/RethinkDb.Driver/blob/master/Source/RethinkDb.Driver/RethinkDb.Driver.project.json) with the `runtimes: { "win-anycpu" }` and `"frameworks": { "net45":{}}`.

This *should* work. But it does not. I think it's likely due to No.1 https://github.com/NuGet/Home/issues/1859. But once https://github.com/NuGet/Home/issues/1859 as you suggested; however, if https://github.com/NuGet/Home/issues/1859 is fixed, I fear I will be in a state where I will no longer be able to build this project. https://github.com/NuGet/Home/issues/1859 actually allows me to build the project normally ***without*** failing on this "runtimes" compiler error.

**Comment by bchavez** [comment]
Please feel free to checkout / clone the repo here:

https://github.com/bchavez/RethinkDb.Driver/tree/master/Source/RethinkDb.Driver

I have MSBuild hacks everywhere just trying to get this to build for `net45` and `dnx` CoreCLR. All the msbuild hacks are inside [this BauBild file](https://github.com/bchavez/RethinkDb.Driver/blob/master/Source/Builder/BauBuild.cs)

**Comment by jasonmalinowski** [comment]
Read the error you're getting carefully: on Update 1, you need `win` as a runtime, not `win-anycpu`. If you change this, your error should go away. The bug I filed is simply to make sure that both of your builds would fail when you're in this broken case, not just one of them.

**Comment by bchavez** [comment]
Ah, very nice. Thank you so very much @jasonmalinowski . Just committed the changes with your suggestion. I no longer need the `*.lock` hack to cross-build. Please give @jasonmalinowski MVP of the year! Thanks :+1: !

**Comment by jasonmalinowski** [comment]
Also, another possible solution: if you really never want your csproj reading any project.json file anywhere, you can just set `<ResolveNuGetPackages>false</ResolveNuGetPackages>` in your .csproj file and that'll turn this all off.

**Comment by bchavez** [comment]
Awesome! There is so much valuable information in this issue undoubtedly someone will find this info very useful. Again, thank you and to everyone for their input in helping unblock this issue.

**Comment by Konard** [comment]
Is there a way not to ignore `project.json` by adding `project-name.project.json`:

```json
{
  "runtimes": { "win": { } },
  "frameworks": { "net45": { } }
}
```

but use 'packages.config' for `csproj` projects? So both can share same code, but each project will have different nuget configuration?

**Comment by Konard** [comment]
> Also, another possible solution: if you really never want your csproj reading any project.json file anywhere, you can just set <ResolveNuGetPackages>false</ResolveNuGetPackages> in your .csproj file and that'll turn this all off.

There is also strange behaiviour if `<ResolveNuGetPackages>false</ResolveNuGetPackages>` is set in `csproj` but `project-name.project.json` removed:

```
------ Discover test started ------
An error occurred while reading file 'C:\Users\Константин\Desktop\LinksPlatform\Platform\Platform.WindowsAPI\Platform.WindowsAPI.project.json': Could not find file 'C:\Users\Константин\Desktop\LinksPlatform\Platform\Platform.WindowsAPI\Platform.WindowsAPI.project.json'.
========== Discover test finished: 0 found (0:00:00,5070015) ==========
```

After this solution is to reload project again, but why it happens in the first place?

**Comment by psulek** [comment]
@bchavez **name.project.json** helps me a lot, thanks!

**Comment by dmitriyse** [comment]
In my case migration to project.json is not an option. I created new issue and suggest solution with temp implementation.  https://github.com/NuGet/Home/issues/3207
